presto-hive/src/main/java/com/facebook/presto/hive/ManifestPartitionLoader.java

```diff
@@ -0,0 +1,175 @@
+    private final Deque<Iterator<InternalHiveSplit>> fileIterators;
+    private final boolean schedulerUsesHostAddresses;
+
+    public ManifestPartitionLoader(
+            Table table,
+            Optional<Domain> pathDomain,
+            ConnectorSession session,
+            HdfsEnvironment hdfsEnvironment,
+            Deque<Iterator<InternalHiveSplit>> fileIterators,
+            boolean schedulerUsesHostAddresses)
+    {
+        this.table = requireNonNull(table, "table is null");
+        this.pathDomain = requireNonNull(pathDomain, "pathDomain is null");
+        this.session = requireNonNull(session, "session is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.hdfsContext = new HdfsContext(session, table.getDatabaseName(), table.getTableName());
+        this.fileIterators = requireNonNull(fileIterators, "fileIterators is null");
+        this.schedulerUsesHostAddresses = schedulerUsesHostAddresses;
+    }
+
+    public ListenableFuture<?> loadPartition(HivePartitionMetadata partition, HiveSplitSource hiveSplitSource, boolean stopped)
+            throws IOException
+    {
+        Path path = new Path(getPartitionLocation(table, partition.getPartition()));
+        Map<String, String> parameters = partition.getPartition().get().getParameters();
+        List<String> fileNames = getFileNames(parameters.get(FILE_NAMES));
+        List<Long> fileSizes = Arrays.stream(parameters.get(FILE_SIZES).split(",")).map(Long::valueOf).collect(toImmutableList());
+
+        // Verify that the count of fileNames and fileSizes are same
+        verify(fileNames.size() == fileSizes.size(), "List of fileNames and fileSizes differ in length");
+
+        ImmutableList.Builder<HiveFileInfo> fileListBuilder = ImmutableList.builder();
+        for (int i = 0; i < fileNames.size(); i++) {
+            Path filePath = new Path(path, fileNames.get(i));
+            FileStatus fileStatus = new FileStatus(fileSizes.get(i), false, 1, getMaxSplitSize(session).toBytes(), 0, filePath);
+            try {
+                BlockLocation[] locations = new BlockLocation[] {new BlockLocation(new String[] {"localhost:50010"}, new String[] {"localhost"}, 0, fileSizes.get(i))};
+                fileListBuilder.add(HiveFileInfo.createHiveFileInfo(new LocatedFileStatus(fileStatus, locations), Optional.empty()));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        InternalHiveSplitFactory splitFactory = createInternalHiveSplitFactory(table, partition, session, pathDomain, hdfsEnvironment, hdfsContext, schedulerUsesHostAddresses);
+
+        fileIterators.addLast(fileListBuilder.build().stream()
+                .map(status -> splitFactory.createInternalHiveSplit(status, true))
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .iterator());
+
```

Comment:
Let's add a new session property to call `directoryLister.list` here to do verification on the file names in manifests are the same as the ones listed.